# Fase 3: Diseño algorítmico y eficiencia computacional

## Proyecto transversal

### Análisis de procedimientos sancionatorios de la Superintendencia del Medio Ambiente (SMA)

---

**Integrantes:**

- Hernán Saavedra González
- Lenin Morales Salas
- Rodrigo Gómez Indellicati


**Objetivo de la fase**

Diseñar, implementar y evaluar estrategias algorítmicas en Python para
el procesamiento y análisis del conjunto de datos de procedimientos
sancionatorios de la SMA, considerando criterios de eficiencia temporal,
uso de memoria, modularidad y reproducibilidad.

**Repositorio del proyecto:**

GitHub - proyecto_sma

## 1. Contexto del análisis

El presente notebook corresponde a la Fase 3 del proyecto transversal y continúa
el trabajo desarrollado durante la Fase 2, donde se realizó la exploración,
limpieza y preparación inicial del conjunto de datos de procedimientos
sancionatorios de la Superintendencia del Medio Ambiente (SMA).

En esta etapa el foco se desplaza desde la preparación de datos hacia el diseño
algorítmico, evaluando estrategias de implementación en Python que permitan
procesar información de manera eficiente, modular y reproducible.

Los algoritmos desarrollados utilizarán como entrada el dataset procesado
obtenido en fases anteriores, manteniendo la separación entre datos originales,
datos transformados y código fuente.

La evaluación considerará criterios de:

- eficiencia temporal;
- consumo de memoria;
- claridad del código;
- reutilización mediante funciones y módulos.

## 2. Fuente de datos y reproducibilidad

El análisis utiliza como fuente el conjunto de datos de procedimientos
sancionatorios publicado por la Superintendencia del Medio Ambiente (SMA).

**Archivo utilizado:**

`Sancionatorios.xlsx`

**Características de la versión utilizada:**

- Fuente: Superintendencia del Medio Ambiente (SMA).
- Fecha de descarga: 09 de septiembre de 2026.
- Hora de descarga: 10:00 hrs.
- Ubicación de la copia original:
  `data/raw/Sancionatorios.xlsx`

La copia almacenada corresponde a la versión original utilizada durante el
proyecto y no debe ser reemplazada por nuevas descargas, debido a que la fuente
puede actualizarse periódicamente.

Para asegurar la reproducibilidad, las etapas posteriores trabajan sobre
versiones procesadas del dataset, manteniendo separación entre:

- datos originales (`data/raw`);
- datos procesados (`data/processed`);
- código fuente (`src`);
- notebooks de análisis.

In [1]:
# Librerías principales

import pandas as pd
import numpy as np

from pathlib import Path
import sys

In [2]:
# Configuración de rutas del proyecto

from pathlib import Path

RUTA_PROCESADO = Path("../data/processed/sancionatorios_procesados.csv")

RUTA_PROCESADO

WindowsPath('../data/processed/sancionatorios_procesados.csv')

In [3]:
# Carga del dataset procesado de Fase 2

df = pd.read_csv(RUTA_PROCESADO)

df.head()

,ProcesoSancionId,Expediente,ProcesoSancionTipoNombre,ProcesoSancionEstado,FechaInicio,FechaTermino,ConfirmaPdC,MultaTotalUTA,LinkSNIFA,UnidadFiscalizableId,Nombre,RegionNombre,ComunaNombre,Latitud,Longitud,CategoriaEconomicaNombre,SubCategoriaEconomicaNombre,LinkSNIFA_UF,FechaActualizacion,DuracionDias
0,2,F-002-2013,Fiscalización,Terminado - Sanción,2013-04-01,2013-08-13,No,94.0,https://snifa.sma.gob.cl/Sancionatorio/Ficha/2...,162,MALL PLAZA EGAÑA - LA REINA,Región Metropolitana,La Reina,-33.453000,-70.570000,Equipamiento,Centro comercial,http://snifa.sma.gob.cl/UnidadFiscalizable/Fic...,2026-09-12,134.0
1,4,D-004-2013,Denuncia,Terminado - Sanción,2013-05-02,2013-10-11,No,25.0,https://snifa.sma.gob.cl/Sancionatorio/Ficha/4...,2488,VERTEDERO DICHAM,Región de los Lagos,Chonchi,-42.627141,-73.841298,Saneamiento Ambiental,Vertedero,http://snifa.sma.gob.cl/UnidadFiscalizable/Fic...,2026-09-12,162.0
2,9,D-003-2013,Denuncia,Terminado - Sanción,2013-03-13,2013-08-19,No,12.0,https://snifa.sma.gob.cl/Sancionatorio/Ficha/9...,10211,ALIMENTOS RILA,Región de Arica y Parinacota,Arica,-18.398400,-70.290600,Agroindustrias,Planta procesadora de productos agrícolas,http://snifa.sma.gob.cl/UnidadFiscalizable/Fic...,2026-09-12,159.0
3,12,D-005-2013,Denuncia,Terminado - Sanción,2013-06-03,2015-03-26,No,1.0,https://snifa.sma.gob.cl/Sancionatorio/Ficha/1...,307,PLANTA PROCESADORA DE PRODUCTOS DEL MAR,Región de Arica y Parinacota,Arica,-18.394700,-70.294500,Pesca y Acuicultura,Elaboración y procesado de recursos hidrobioló...,http://snifa.sma.gob.cl/UnidadFiscalizable/Fic...,2026-09-12,661.0
4,13,F-007-2013,Fiscalización,Terminado - Sanción,2013-06-03,2013-11-27,No,865.0,https://snifa.sma.gob.cl/Sancionatorio/Ficha/1...,2435,REFINERIA ENAP BIOBIO,Región del Biobío,Hualpén,-36.775327,-73.116842,Instalación fabril,Planta de Refinería,http://snifa.sma.gob.cl/UnidadFiscalizable/Fic...,2026-09-12,177.0


In [4]:
# Dimensiones del dataset de entrada

df.shape

(984, 20)

In [5]:
# Información general del dataset

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 984 entries, 0 to 983
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ProcesoSancionId             984 non-null    int64  
 1   Expediente                   984 non-null    str    
 2   ProcesoSancionTipoNombre     984 non-null    str    
 3   ProcesoSancionEstado         984 non-null    str    
 4   FechaInicio                  984 non-null    str    
 5   FechaTermino                 982 non-null    str    
 6   ConfirmaPdC                  984 non-null    str    
 7   MultaTotalUTA                984 non-null    float64
 8   LinkSNIFA                    984 non-null    str    
 9   UnidadFiscalizableId         984 non-null    int64  
 10  Nombre                       984 non-null    str    
 11  RegionNombre                 984 non-null    str    
 12  ComunaNombre                 984 non-null    str    
 13  Latitud                      97

## 3. Diseño algorítmico propuesto

A partir del conjunto de datos procesado en la Fase 2, se diseñan algoritmos orientados a generar información analítica sobre los procedimientos sancionatorios de la Superintendencia del Medio Ambiente (SMA).

El diseño considera una separación entre la lógica de procesamiento y la visualización de resultados, utilizando funciones reutilizables almacenadas en módulos Python dentro de la carpeta `src/`.

Considerando las características del dataset de entrada, las decisiones algorítmicas se orientan a construir soluciones modulares, reproducibles y eficientes.

El problema no presenta una alta magnitud de registros, por lo que la mejora computacional se enfocará principalmente en:

- organización del procesamiento;
- reutilización de funciones;
- reducción de operaciones repetidas;
- claridad y mantenimiento del código.

Se evaluarán dos estrategias de implementación:

## Programación estructurada

La estrategia principal utilizada en el proyecto corresponde a programación estructurada mediante funciones independientes, donde cada función representa una operación específica del procesamiento analítico.

Este enfoque permite separar responsabilidades y facilitar la reutilización del código. Las principales operaciones consideradas son:

- construcción del universo analítico;
- cálculo de indicadores asociados a procedimientos sancionatorios;
- cálculo de duración de procesos;
- normalización y transformación de variables.

La implementación mediante funciones permite que el notebook actúe como controlador del flujo de análisis, mientras que la lógica de procesamiento permanece encapsulada dentro de módulos Python ubicados en la carpeta `src/`.

Desde el punto de vista de eficiencia temporal, se priorizan operaciones vectorizadas mediante Pandas, evitando iteraciones manuales innecesarias sobre los registros del DataFrame.

## Programación recursiva

Las funciones recursivas son una alternativa útil cuando el problema presenta una estructura jerárquica o requiere dividir un problema en subproblemas equivalentes.

Sin embargo, para el análisis de procedimientos sancionatorios de la SMA, los datos presentan una estructura tabular donde predominan operaciones de agrupación, transformación y cálculo de indicadores.

Por esta razón, se evaluará críticamente su aplicación, considerando que una solución recursiva podría aumentar el consumo de memoria debido a la utilización de la pila de llamadas, sin entregar ventajas significativas frente a las operaciones vectorizadas disponibles en Pandas.

La selección del enfoque algorítmico se realizará considerando:

- tiempo de ejecución;
- consumo de memoria;
- claridad del código;
- facilidad de mantenimiento.

# 4. Implementación modular de algoritmos

Para mejorar la mantenibilidad y reproducibilidad del proyecto, la lógica de procesamiento se separa del notebook principal mediante módulos Python ubicados en la carpeta `src/`.

Esta organización permite que el notebook actúe como controlador del flujo de análisis, mientras que las funciones de procesamiento pueden ser reutilizadas, probadas y versionadas de manera independiente.

La estructura implementada es:

src/
├── procesamiento.py
├── algoritmos.py
└── mediciones.py

Cada módulo tiene una responsabilidad específica:

- `procesamiento.py`: contiene funciones relacionadas con preparación y transformación de datos.
- `algoritmos.py`: contiene funciones orientadas al análisis y generación de indicadores.
- `mediciones.py`: contiene funciones para evaluar tiempo de ejecución y eficiencia computacional.

Esta separación facilita la trazabilidad del código y permite evaluar el desempeño de los algoritmos desarrollados.

In [8]:
from pathlib import Path
import sys

# Obtener raíz del proyecto
RAIZ_PROYECTO = Path.cwd().parent

# Agregar raíz al path de Python
sys.path.append(str(RAIZ_PROYECTO))

RAIZ_PROYECTO

WindowsPath('C:/Users/hernan.saavedra/Documents/proyecto_sma')

In [9]:
from src.procesamiento import construir_universo_analitico
from src.procesamiento import calcular_duracion_proceso

In [10]:
universo = construir_universo_analitico(df)

universo.head()

,UnidadFiscalizableId,cantidad_procesos,multa_total
0,155,1,9.5
1,156,1,139.2
2,162,2,235.0
3,220,1,304.0
4,227,1,60.0


### Resultado del algoritmo: construcción del universo analítico

La función `construir_universo_analitico()` permite transformar el dataset de procedimientos sancionatorios en una estructura resumida por unidad fiscalizable.

El algoritmo utiliza operaciones de agrupación mediante Pandas (`groupby`), permitiendo obtener indicadores agregados como:

- cantidad de procedimientos asociados a una unidad fiscalizable;
- monto total acumulado de multas.

La estrategia utilizada corresponde a un enfoque estructurado y vectorizado, evitando recorrer manualmente cada registro mediante ciclos, lo que mejora la legibilidad y aprovecha las optimizaciones internas de la biblioteca Pandas.

Desde el punto de vista de complejidad, la operación presenta un comportamiento aproximado O(n), debido a que requiere procesar los registros del DataFrame de entrada para generar los agrupamientos correspondientes.

In [11]:
datos_duracion = calcular_duracion_proceso(df)

datos_duracion[
    ["ProcesoSancionId", "FechaInicio", "FechaTermino", "DuracionDias"]
].head()

,ProcesoSancionId,FechaInicio,FechaTermino,DuracionDias
0,2,2013-04-01,2013-08-13,134.0
1,4,2013-05-02,2013-10-11,162.0
2,9,2013-03-13,2013-08-19,159.0
3,12,2013-06-03,2015-03-26,661.0
4,13,2013-06-03,2013-11-27,177.0


### Resultado del algoritmo: cálculo de duración de procedimientos

La función `calcular_duracion_proceso()` permite obtener la duración temporal de cada procedimiento sancionatorio mediante la diferencia entre la fecha de inicio y la fecha de término.

Este procesamiento genera una nueva variable denominada `DuracionDias`, la cual permite analizar posteriormente el comportamiento temporal de los procesos sancionatorios.

La implementación utiliza operaciones vectorizadas de Pandas para transformar las fechas y calcular diferencias entre columnas, evitando iteraciones registro por registro.

Desde la perspectiva de eficiencia algorítmica, la operación presenta una complejidad aproximada O(n), debido a que cada registro del conjunto de datos es procesado una única vez.

El uso de funciones independientes favorece la modularidad del proyecto, permitiendo reutilizar este cálculo en diferentes etapas del análisis sin duplicar código.

## 5. Medición de eficiencia computacional

Para evaluar empíricamente el comportamiento de los algoritmos desarrollados, se incorporará una medición del tiempo de ejecución.

La medición se realizará utilizando `time.perf_counter()`, que permite registrar el tiempo transcurrido durante la ejecución de una función.

El objetivo es complementar el análisis teórico de complejidad con una medición observable sobre el conjunto de datos utilizado en el proyecto.

In [12]:
from src.mediciones import medir_tiempo

In [13]:
universo, tiempo_universo = medir_tiempo(
    construir_universo_analitico,
    df
)

print(f"Tiempo de ejecución: {tiempo_universo:.6f} segundos")

Tiempo de ejecución: 0.009714 segundos


In [14]:
datos_duracion, tiempo_duracion = medir_tiempo(
    calcular_duracion_proceso,
    df
)

print(f"Tiempo de ejecución: {tiempo_duracion:.6f} segundos")

Tiempo de ejecución: 0.005880 segundos


## 5.1. Comparación de tiempos de ejecución

Las mediciones realizadas sobre el conjunto de datos utilizado muestran los siguientes tiempos de ejecución:

- `construir_universo_analitico()`: 0.009714 segundos.
- `calcular_duracion_proceso()`: 0.005880 segundos.

La diferencia observada se relaciona con las operaciones que realiza cada algoritmo. La construcción del universo analítico requiere agrupar los registros mediante `groupby()` y calcular indicadores agregados, mientras que el cálculo de duración realiza principalmente transformaciones de fechas y una resta entre columnas.

Estos valores corresponden a una ejecución puntual sobre un dataset de 984 registros, por lo que no deben interpretarse como una medición definitiva del rendimiento general de los algoritmos. Su utilidad en esta fase es complementar el análisis teórico de complejidad con evidencia experimental.

Ambos procedimientos mantienen una complejidad aproximada O(n), ya que el procesamiento depende del número de registros de entrada. La medición empírica permite observar el costo de ejecución para el tamaño actual del dataset y establecer una línea base para futuras comparaciones.

In [15]:
import time

def medir_repetidamente(funcion, datos, repeticiones=10):
    tiempos = []

    for _ in range(repeticiones):
        inicio = time.perf_counter()
        funcion(datos)
        fin = time.perf_counter()

        tiempos.append(fin - inicio)

    return {
        "minimo": min(tiempos),
        "maximo": max(tiempos),
        "promedio": sum(tiempos) / len(tiempos)
    }

In [16]:
resultado_universo = medir_repetidamente(
    construir_universo_analitico,
    df
)

resultado_universo

{'minimo': 0.004515399981755763,
 'maximo': 0.006139300006907433,
 'promedio': 0.005430669995257631}

In [17]:
resultado_duracion = medir_repetidamente(
    calcular_duracion_proceso,
    df
)

resultado_duracion

{'minimo': 0.00397970003541559,
 'maximo': 0.0052921999595128,
 'promedio': 0.004733159992611036}

In [18]:
comparacion_tiempos = pd.DataFrame({
    "Algoritmo": [
        "Construcción del universo analítico",
        "Cálculo de duración de procesos"
    ],
    "Tiempo mínimo (s)": [
        resultado_universo["minimo"],
        resultado_duracion["minimo"]
    ],
    "Tiempo máximo (s)": [
        resultado_universo["maximo"],
        resultado_duracion["maximo"]
    ],
    "Tiempo promedio (s)": [
        resultado_universo["promedio"],
        resultado_duracion["promedio"]
    ]
})

comparacion_tiempos

,Algoritmo,Tiempo mínimo (s),Tiempo máximo (s),Tiempo promedio (s)
0,Construcción del universo analítico,0.004515,0.006139,0.005431
1,Cálculo de duración de procesos,0.003980,0.005292,0.004733


## 5.2. Análisis de los resultados

Las mediciones repetidas permiten complementar el análisis teórico de complejidad con evidencia experimental.

Para la construcción del universo analítico se obtuvo un tiempo promedio de 0,005431 segundos, mientras que el cálculo de duración de los procesos presentó un promedio de 0,004733 segundos.

La diferencia observada es pequeña y debe interpretarse considerando que ambos algoritmos realizan operaciones diferentes. La construcción del universo analítico utiliza agrupación mediante `groupby()` y operaciones de agregación, mientras que el cálculo de duración realiza transformaciones de fechas y operaciones vectorizadas.

En ambos casos, el procesamiento depende del número de registros de entrada, por lo que se considera una complejidad temporal aproximada O(n). Los tiempos obtenidos corresponden al dataset actual de 984 registros y constituyen una línea base para futuras mediciones.

La repetición de las pruebas permite reducir la dependencia de una única ejecución y proporciona una referencia más estable para comparar modificaciones posteriores del código.

In [19]:
import tracemalloc


def medir_memoria(funcion, datos):
    tracemalloc.start()

    resultado = funcion(datos)

    memoria_actual, memoria_pico = tracemalloc.get_traced_memory()

    tracemalloc.stop()

    return resultado, memoria_pico

In [20]:
universo_memoria, memoria_universo = medir_memoria(
    construir_universo_analitico,
    df
)

print(f"Memoria máxima utilizada: {memoria_universo / 1024:.2f} KB")

Memoria máxima utilizada: 74.83 KB


In [21]:
duracion_memoria, memoria_duracion = medir_memoria(
    calcular_duracion_proceso,
    df
)

print(f"Memoria máxima utilizada: {memoria_duracion / 1024:.2f} KB")

Memoria máxima utilizada: 231.03 KB


In [22]:
comparacion_eficiencia = pd.DataFrame({
    "Algoritmo": [
        "Construcción del universo analítico",
        "Cálculo de duración de procesos"
    ],
    "Tiempo promedio (s)": [
        resultado_universo["promedio"],
        resultado_duracion["promedio"]
    ],
    "Memoria máxima (KB)": [
        memoria_universo / 1024,
        memoria_duracion / 1024
    ]
})

comparacion_eficiencia

,Algoritmo,Tiempo promedio (s),Memoria máxima (KB)
0,Construcción del universo analítico,0.005431,74.832031
1,Cálculo de duración de procesos,0.004733,231.031250


### 5.3. Análisis conjunto de tiempo y memoria

La comparación permite observar que ambos algoritmos presentan tiempos de ejecución bajos para el conjunto de datos utilizado. El cálculo de duración presentó un tiempo promedio de 0,004733 segundos, mientras que la construcción del universo analítico presentó un promedio de 0,005431 segundos.

En términos de memoria, se observa un comportamiento diferente. La construcción del universo analítico registró un uso máximo aproximado de 74,83 KB, mientras que el cálculo de duración alcanzó aproximadamente 231,03 KB.

Esta diferencia se relaciona con la implementación utilizada. La función `calcular_duracion_proceso()` crea una copia del DataFrame y realiza transformaciones sobre las columnas de fecha, generando estructuras temporales adicionales. Por lo tanto, un menor tiempo de ejecución no implica necesariamente un menor consumo de memoria.

Para ambos algoritmos se considera un comportamiento temporal aproximado O(n), dado que el procesamiento depende del número de registros de entrada. El consumo de memoria también depende del tamaño de los datos y de las estructuras temporales generadas durante las operaciones.

Los resultados obtenidos constituyen una línea base experimental para comparar futuras mejoras de implementación.

## 5.4. Comparación entre programación estructurada y recursiva

Para el procesamiento desarrollado en esta fase se prioriza la programación estructurada mediante funciones independientes y operaciones vectorizadas de Pandas. Este enfoque resulta adecuado para trabajar con un DataFrame, ya que permite separar cada tarea en funciones específicas y mantener un flujo de procesamiento claro.

La programación recursiva puede ser útil cuando el problema presenta una estructura jerárquica o cuando una solución puede expresarse naturalmente como un problema que se reduce a instancias más pequeñas del mismo problema. Sin embargo, en el conjunto de datos analizado no se identifica una estructura jerárquica que justifique reemplazar las operaciones vectorizadas por una solución recursiva.

Una implementación recursiva también implica considerar el consumo de memoria asociado a las llamadas de función y la profundidad de la recursión. Por este motivo, utilizar recursividad solamente por el hecho de que forma parte de los contenidos de la fase no necesariamente representa una mejora de eficiencia.

En este proyecto, la decisión es mantener la programación estructurada para las transformaciones y agregaciones del DataFrame, y utilizar la recursividad como estrategia a evaluar en problemas donde su estructura resulte adecuada. La elección se fundamenta en la correspondencia entre la naturaleza del problema, la claridad del código y los costos de tiempo y memoria observados.

In [23]:
def suma_iterativa(n):
    total = 0

    for i in range(1, n + 1):
        total += i

    return total


def suma_recursiva(n):
    if n == 0:
        return 0

    return n + suma_recursiva(n - 1)

In [24]:
n = 100

resultado_iterativo = suma_iterativa(n)
resultado_recursivo = suma_recursiva(n)

print("Resultado iterativo:", resultado_iterativo)
print("Resultado recursivo:", resultado_recursivo)

Resultado iterativo: 5050
Resultado recursivo: 5050


In [25]:
_, tiempo_iterativo = medir_tiempo(suma_iterativa, n)
_, tiempo_recursivo = medir_tiempo(suma_recursiva, n)

print(f"Tiempo iterativo: {tiempo_iterativo:.8f} segundos")
print(f"Tiempo recursivo: {tiempo_recursivo:.8f} segundos")

Tiempo iterativo: 0.00000680 segundos
Tiempo recursivo: 0.00001120 segundos


In [26]:
resultado_iterativo = medir_repetidamente(
    suma_iterativa,
    n
)

resultado_recursivo = medir_repetidamente(
    suma_recursiva,
    n
)

print("Iterativo:")
print(resultado_iterativo)

print("\nRecursivo:")
print(resultado_recursivo)

Iterativo:
{'minimo': 4.00003045797348e-06, 'maximo': 6.200047209858894e-06, 'promedio': 4.390004323795438e-06}

Recursivo:
{'minimo': 8.100003469735384e-06, 'maximo': 1.1899974197149277e-05, 'promedio': 8.58000130392611e-06}


In [27]:
_, memoria_iterativo = medir_memoria(
    suma_iterativa,
    n
)

_, memoria_recursivo = medir_memoria(
    suma_recursiva,
    n
)

print(f"Memoria iterativa: {memoria_iterativo / 1024:.2f} KB")
print(f"Memoria recursiva: {memoria_recursivo / 1024:.2f} KB")

Memoria iterativa: 0.00 KB
Memoria recursiva: 0.00 KB


In [28]:
print(f"Memoria iterativa: {memoria_iterativo} bytes")
print(f"Memoria recursiva: {memoria_recursivo} bytes")

Memoria iterativa: 0 bytes
Memoria recursiva: 0 bytes


## 5.5. Evaluación experimental de los enfoques iterativo y recursivo

Para complementar la comparación conceptual, se realizó una prueba experimental utilizando una función iterativa y una función recursiva para resolver el mismo problema de suma de valores.

Ambas implementaciones fueron ejecutadas sobre el mismo valor de entrada y posteriormente se compararon sus resultados, tiempos de ejecución y comportamiento en términos de memoria.

## 5.5.1. Comparación de los enfoques iterativo y recursivo

La prueba controlada permitió comprobar que ambas implementaciones producen el mismo resultado. Para `n = 100`, la suma obtenida fue 5050 en ambos casos.

Las mediciones repetidas mostraron un tiempo promedio de 0,00000439 segundos para la implementación iterativa y de 0,00000858 segundos para la implementación recursiva.

En la medición de memoria, `tracemalloc` no registró asignaciones observables para este problema de pequeña escala, por lo que estos resultados no permiten establecer una diferencia experimental de consumo de memoria. Por este motivo, se complementa la medición con el análisis teórico de complejidad espacial.

La solución iterativa utiliza una cantidad constante de memoria auxiliar, por lo que presenta un comportamiento O(1) en espacio. La solución recursiva requiere mantener las llamadas activas en la pila de ejecución, por lo que su consumo de espacio crece con el número de llamadas y presenta un comportamiento O(n).

En este ejemplo, la implementación recursiva presenta un mayor tiempo promedio y además requiere espacio adicional asociado a la profundidad de la recursión. Estos resultados sirven como referencia para comprender las diferencias entre ambos enfoques, pero no se generalizan directamente a todos los problemas.

In [31]:
def contar_por_unidad_iterativo(df):
    conteo = {}

    for unidad in df["UnidadFiscalizableId"]:
        if unidad in conteo:
            conteo[unidad] += 1
        else:
            conteo[unidad] = 1

    return conteo

In [32]:
def contar_por_unidad_pandas(df):
    return df["UnidadFiscalizableId"].value_counts().to_dict()

In [33]:
resultado_iterativo = contar_por_unidad_iterativo(df)
resultado_pandas = contar_por_unidad_pandas(df)

print("Primeros resultados - versión iterativa:")
print(list(resultado_iterativo.items())[:5])

print("\nPrimeros resultados - versión Pandas:")
print(list(resultado_pandas.items())[:5])

Primeros resultados - versión iterativa:
[(162, 2), (2488, 2), (10211, 1), (307, 1), (2435, 1)]

Primeros resultados - versión Pandas:
[(1381, 3), (13738, 3), (13957, 3), (162, 2), (2488, 2)]


In [34]:
assert resultado_iterativo == resultado_pandas

print("✓ Ambas implementaciones producen el mismo resultado.")

✓ Ambas implementaciones producen el mismo resultado.


In [35]:
resultado_tiempo_iterativo = medir_repetidamente(
    contar_por_unidad_iterativo,
    df
)

resultado_tiempo_pandas = medir_repetidamente(
    contar_por_unidad_pandas,
    df
)

print("Versión iterativa:")
print(resultado_tiempo_iterativo)

print("\nVersión Pandas:")
print(resultado_tiempo_pandas)

Versión iterativa:
{'minimo': 0.0002949000336229801, 'maximo': 0.0004091999726369977, 'promedio': 0.0003305099962744862}

Versión Pandas:
{'minimo': 0.00048480002442374825, 'maximo': 0.0012164999498054385, 'promedio': 0.0007933499931823462}


In [36]:
_, memoria_iterativa_unidad = medir_memoria(
    contar_por_unidad_iterativo,
    df
)

_, memoria_pandas_unidad = medir_memoria(
    contar_por_unidad_pandas,
    df
)

print(
    f"Memoria versión iterativa: "
    f"{memoria_iterativa_unidad / 1024:.2f} KB"
)

print(
    f"Memoria versión Pandas: "
    f"{memoria_pandas_unidad / 1024:.2f} KB"
)

Memoria versión iterativa: 72.36 KB
Memoria versión Pandas: 91.75 KB


## 5.6 Comparación de implementaciones

Para evaluar la eficiencia de dos estrategias de implementación, se comparó una solución iterativa mediante `for` con una solución basada en `value_counts()` de Pandas. Ambas implementaciones fueron aplicadas al mismo conjunto de datos y se verificó previamente que producen resultados equivalentes.

| Implementación | Tiempo promedio (s) | Memoria máxima (KB) |
|---|---:|---:|
| Iterativa (`for`) | 0.000331 | 72.36 |
| Pandas (`value_counts`) | 0.000793 | 91.75 |

Los resultados muestran que, para el conjunto de datos utilizado, la implementación iterativa presentó un menor tiempo promedio de ejecución y un menor consumo máximo de memoria.

Esta diferencia corresponde a las condiciones específicas de la prueba y al tamaño del dataset analizado. Por lo tanto, los resultados se interpretan como una medición experimental del caso estudiado y no como una conclusión general sobre la eficiencia de ambas estrategias.

La equivalencia funcional fue comprobada mediante una comparación de los resultados obtenidos por ambas implementaciones.

# 6. Conclusiones

La Fase 3 permitió diseñar, implementar y evaluar distintas estrategias algorítmicas aplicadas al conjunto de datos de procedimientos sancionatorios de la Superintendencia del Medio Ambiente.

A partir de la implementación de funciones modulares en Python, se logró separar la lógica de procesamiento, los algoritmos y las funciones destinadas a medir su eficiencia. Esta organización favorece la reutilización, trazabilidad y mantenimiento del código.

Las mediciones experimentales permitieron complementar el análisis teórico de complejidad. Se observaron diferencias en los tiempos de ejecución y en el consumo de memoria dependiendo de las operaciones realizadas y de la estrategia de implementación utilizada.

La comparación entre una implementación iterativa y una implementación basada en Pandas permitió comprobar que ambas estrategias pueden producir el mismo resultado presentando, para el conjunto de datos analizado, diferentes comportamientos en tiempo y memoria.

Asimismo, la comparación entre programación iterativa y recursiva permitió observar las diferencias asociadas al uso de la pila de ejecución y al costo temporal de las llamadas recursivas. Estos resultados deben interpretarse considerando las características específicas del problema y del conjunto de datos utilizado.

En conjunto, las experiencias desarrolladas permiten establecer una línea base experimental para futuras mejoras del proyecto y muestran la importancia de considerar simultáneamente corrección, complejidad, tiempo de ejecución, consumo de memoria, modularidad y reproducibilidad al seleccionar una estrategia algorítmica.